nodestream_akamai/akamai_utils/property_client.py

```diff
@@ -222,292 +300,163 @@ def list_all_properties(self):
 
         # Extract origin behaviour itself and append hostname to list based on origin type
         origin_behavior_match = parse(origin_location)
         origin_search = origin_behavior_match.find(rules)
         origin_behavior = origin_search[0].value
-        if "hostname" in origin_behavior["options"].keys():
+        if "hostname" in origin_behavior["options"]:
             origin_host = origin_behavior["options"]["hostname"]
-        elif "netStorage" in origin_behavior["options"].keys():
+        elif "netStorage" in origin_behavior["options"]:
             origin_host = origin_behavior["options"]["netStorage"]["downloadDomainName"]
-        elif "mslorigin" in origin_behavior["options"].keys():
+        elif "mslorigin" in origin_behavior["options"]:
             origin_host = origin_behavior["options"]["mslorigin"]
 
         # Split JSONPATH into children[X] elements so we can iterate down the path
-        location_elements = re.findall("children\.\[[\d]+\]", rule_base)
+        location_elements = re.findall(r"children\.\[\d+]", rule_base)
         return origin_host, location_elements
 
-    def parse_origin_location(self, rule_location, parent_location, rules):
+    def parse_origin_location(self, rule_location_input, parent_location, rules):
         """
         This function parses origin locations to extract path matches, hostname matches
         and Conditional Origin IDs
         """
+        self.logger.debug(
+            "parse_origin_location(rule_location=%s, parent_location=%s)",
+            rule_location_input,
+            parent_location,
+        )
         # Construct rule location from element and optionally parent path
         if parent_location != "":
-            rule_location = parent_location + "." + rule_location
+            rule_location = parent_location + "." + rule_location_input
+        else:
+            rule_location = rule_location_input
+
         rule_match = parse(rule_location)
         rule_search = rule_match.find(rules)
 
-        if len(rule_search) == 0:
-            raise (Exception(f"No rule found at position '{rule_location}'"))
-
-        # Define criteria to look at
-        match_types = ["path", "hostname", "cloudletsOrigin"]
-
-        # Define negative criterion matches
-        negative_operators = ["DOES_NOT_MATCH_ONE_OF", "IS_NOT_ONE_OF"]
+        if rule_search is None or len(rule_search) == 0:
+            self.logger.warning("No rule found at position '%s'", rule_location)
+            return {}
 
         # Extract rule by JSONPATH
         rule = rule_search[0].value
 
         # Instantiate results
         criteria_results = {}
-        rule_results = {}
+        rule_results = {k: [] for k in MATCH_TYPES}
         location_results = {}
 
-        for match_type in match_types:
+        for match_type in MATCH_TYPES:
             criteria_results[match_type] = []
-            criterion_results = {}
+            criterion_results = {k: [] for k in MATCH_TYPES}

```

Comment:
sometimes the inner loops were trying to use the keys before they were added to the dict